compiler/src/dotty/tools/backend/sjs/JSExportsGen.scala

```diff
@@ -553,10 +544,25 @@ final class JSExportsGen(jsCodeGen: JSCodeGen)(using Context) {
 
       if (altsByTypeTest.size == 1) {
         // Testing this parameter is not doing any us good
-        genExportSameArgcRec(jsName, formalArgsRegistry, alts, paramIndex + 1, static, maxArgc)
+        genOverloadDispatchSameArgcRec(jsName, formalArgsRegistry, alts, tpe, paramIndex + 1, maxArgc)
       } else {
         // Sort them so that, e.g., isInstanceOf[String] comes before isInstanceOf[Object]
-        val sortedAltsByTypeTest = topoSortDistinctsBy(altsByTypeTest)(_._1)
+        val sortedAltsByTypeTest = topoSortDistinctsWith(altsByTypeTest) { (lhs, rhs) =>
+          (lhs._1, rhs._1) match {
+            // NoTypeTest is always last
+            case (_, NoTypeTest) => true
+            case (NoTypeTest, _) => false
+
+            case (PrimitiveTypeTest(_, rank1), PrimitiveTypeTest(_, rank2)) =>
+              rank1 <= rank2
+
+            case (InstanceOfTypeTest(t1), InstanceOfTypeTest(t2)) =>
+              t1 <:< t2
```

Comment:
Aside, but you can turn a partial ordering based on subtyping into a total ordering by replacing it with a check for the length of the base type sequence (https://github.com/lampepfl/dotty/commit/171cea2127eed5c54691a3984c2c1546c86f444e), which should allow you to use a regular sort function instead of doing a topological sort.

Reply:
I don't think that's true. Given
```scala
class A
class B extends A
class C extends A
```
`B` and `C` have the same base type sequence length, yet they are not the same. That's not a total ordering. Perhaps it works in 171cea2127eed5c54691a3984c2c1546c86f444e because it's known to be in the same superclass lineage? But in this case we don't know that.

Reply:
> That's not a total ordering. 

It gives you a total ordering for equivalence classes, where two types with the same base length are considered to be equivalent which is good enough for our purposes (I've also recently learned this is called a strict weak ordering: https://en.wikipedia.org/wiki/Weak_ordering#Strict_weak_orderings)